src/preprocessing/passes/bool_to_bv.cpp

```diff
@@ -221,7 +215,157 @@ void BoolToBV::lowerNode(const TNode& n)
     default: break;
   }
 
+  // check if it's safe to lower or rebuild the node
+  // Note: might have to rebuild to keep changes to children, even if this node
+  // isn't being lowered
+
+  // it's safe to lower if all the children are bit-vectors
+  bool safe_to_lower =
+      (new_kind != k);  // don't need to lower at all if kind hasn't changed
+
+  // it's safe to rebuild if rebuilding doesn't change any of the types of the
+  // children
+  bool safe_to_rebuild = true;
+
+  for (const Node& nn : n)
+  {
+    if (safe_to_lower)
+    {
+      safe_to_lower = fromCache(nn).getType().isBitVector();
+    }
+    if (safe_to_rebuild)
+    {
+      safe_to_rebuild = (fromCache(nn).getType() == nn.getType());
+    }
+    // if it's already not safe to do either, stop checking
+    if (!safe_to_lower && !safe_to_rebuild)
+    {
+      break;
+    }
+  }
+
+  Debug("bool-to-bv") << "safe_to_lower = " << safe_to_lower
+                      << ", safe_to_rebuild = " << safe_to_rebuild << std::endl;
+
+  if (new_kind != k)
+  {
+    if (safe_to_lower)
+    {
+      rebuildNode(n, new_kind);
+      return;
+    }
+
+    else if (force && fromCache(n).getType().isBoolean())
+    {
+      if (safe_to_rebuild && needToRebuild(n))
+      {
+        // need to rebuild to keep changes made to descendants
+        rebuildNode(n, k);
+      }
+
+      updateCache(n,
+                  nm->mkNode(kind::ITE,
+                             fromCache(n),
+                             bv::utils::mkOne(1),
+                             bv::utils::mkZero(1)));
+      Debug("bool-to-bv") << "BoolToBV::lowerNodeHelper forcing " << n
+                          << " =>\n"
+                          << fromCache(n) << std::endl;
+      ++(d_statistics.d_numTermsForcedLowered);
+      return;
+    }
+  }
+  else if (safe_to_rebuild && needToRebuild(n))
+  {
+    Assert(k == new_kind);
+    rebuildNode(n, k);
+  }
+  else if (force && fromCache(n).getType().isBoolean())
+  {
+    // force booleans (which haven't already been converted) to bit-vector
+    // needed to maintain the invariant that all boolean children
+    // have been converted (even constants and variables) when forcing
+    updateCache(
+        n, nm->mkNode(kind::ITE, n, bv::utils::mkOne(1), bv::utils::mkZero(1)));
+    Debug("bool-to-bv") << "BoolToBV::lowerNodeHelper forcing " << n
+                        << " =>\n"
+                        << fromCache(n) << std::endl;
+    ++(d_statistics.d_numTermsForcedLowered);
+  }
+  else
+  {
+    Debug("bool-to-bv") << "BoolToBV::lowerNodeHelper skipping: " << n
+                        << std::endl;
+  }
+}
+
+Node BoolToBV::lowerIte(const TNode& node)
+{
+  std::vector<TNode> visit;
+  visit.push_back(node);
+  std::unordered_set<TNode, TNodeHashFunction> visited;
```

Comment:
(optional) Perhaps `childrenEnqueued`?

Reply:
`visited` is fine, that's the pattern we commonly use.

Reply:
Left it the same.